### PR TITLE
update cri-plugin to parse hugepages limit

### PR DESCRIPTION
**Pull request by bg-chun**
### The purpose of this PR
- updating CRI vendor to refer the last CRI Update which includes hugepages update
- updating cri-plugin(this repo) to parse newly added hugepages fields from CRI message.

### History
- In Kubernetes, hugepages are requested as the container level resources, but limit is set on only pod level cgroup due to the lack of CRI support.
- But now, CRI has been updated to have hugepages field so that kubernetes be able to set hugepages limit on container level cgroup.
- The last step we have to do is updating container runtime which handles container level cgroup to parse new hugepage field from CRI message and set limit on container level cgroup.

### The Detail of CRI Update
- Now, LinuxContainerResources CRI message, which has linux resource fileds that equals to linux cgroup subsystems , has hugepage field(`hugepage_limits`), it is able to express list of hugepage configuration.
- New hugepage fields are written based on existing hugepage fields of [OCI runtime spec](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#huge-page-limits) so that is is easy to transfrom hugepages fields and values from CRI message to OCI Runtime configuration.
- Each hugepage configuration(`HugepageLimit`) expesss hugepage limit per page size.
- The configuration [page_size:  "2MB", limit: 209715200] means set 2GB limits for 2MB hugepage size on the container level cgroup sandbox.
- More details are available on this [slide](https://docs.google.com/presentation/d/1J_6RGeNU1f21TchLrE672SfFLcbhPUev356NLGYxoUU/edit#slide=id.g7a7421ffe7_0_157) and [doc](https://docs.google.com/document/d/1Xnb3ioEH_JmWLdFJ-i6Lp7jFc_fDuzzAa2oEXJglBbM/edit#heading=h.va9fkr521iz7).

```
message LinuxContainerResources {
    ...
    string cpuset_mems = 7;
    // List of HugepageLimits to limit the HugeTLB usage of container per page size. Default: nil (not specified).
    repeated HugepageLimit hugepage_limits = 8;
}
// HugepageLimit corresponds to the file`hugetlb.<hugepagesize>.limit_in_byte` in container level cgroup.
// For example, `PageSize=1GB`, `Limit=1073741824` means setting `1073741824` bytes to hugetlb.1GB.limit_in_bytes.
message HugepageLimit {
    // The value of PageSize has the format <size><unit-prefix>B (2MB, 1GB),
    // and must match the <hugepagesize> of the corresponding control file found in `hugetlb.<hugepagesize>.limit_in_bytes`.
    // The values of <unit-prefix> are intended to be parsed using base 1024("1KB" = 1024, "1MB" = 1048576, etc).
    string page_size = 1;
    // limit in bytes of hugepagesize HugeTLB usage.
    uint64 limit = 2;
}
```
- KEP updates : https://github.com/kubernetes/enhancements/pull/1199
- CRI Updates : https://github.com/kubernetes/kubernetes/pull/83614

### CODE REVIEW

File: pkg/containerd/opts/spec_unix.go [hunk 1]
```diff
@@ -483,6 +484,14 @@ func WithResources(resources *runtime.LinuxContainerResources) oci.SpecOpts {
 		if limit != 0 {
 			s.Linux.Resources.Memory.Limit = &limit
 		}
+		if len(hugepages) != 0 {
```
Comment (jterry75) [inline pkg/containerd/opts/spec_unix.go]: Why the extra check? `rage` handles `nil` and `len() == 0` and will simply not enumerate

**Comment by mikebrow** [comment]
```
I1104 20:49:04.816] golangci-lint run
I1104 20:49:42.591] pkg/containerd/opts/spec_unix.go:462: File is not `gofmt`-ed with `-s` (gofmt)
I1104 20:49:42.592] 			p      = uint64(resources.GetCpuPeriod())
I1104 20:49:42.592] 			q      = resources.GetCpuQuota()
I1104 20:49:42.592] 			shares = uint64(resources.GetCpuShares())
I1104 20:49:42.594] 			limit  = resources.GetMemoryLimitInBytes()
W1104 20:50:08.701] make: *** [Makefile:58: lint] Error 1
```

**Comment by Random-Liu** [comment]
@bg-chun Can you link to the KEP in the first commit? So that we can better track features. :)

File: pkg/containerd/opts/spec_unix.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/containerd/opts/spec_unix.go]: Agree.

**Comment by bg-chun** [comment]
I will update pr after CRI pr merged on kubernetes.

File: pkg/containerd/opts/spec_unix.go (hunk 1 above)
Comment (bg-chun) [reply pkg/containerd/opts/spec_unix.go]: done

**Comment by bg-chun** [comment]
As temporary, I updated cri-api vendor to point master branch.
I think we will may have new alpha release tag for (1.18.0-alpha.1) soon.
When kubernetes sig-rel cut new tag I will update the vendor from master to new tag.
https://github.com/kubernetes/sig-release/issues/927

**Comment by bg-chun** [comment]
> 
> 
> @bg-chun Can you link to the KEP in the first commit? So that we can better track features. :)
I updated first commit to mention KEP and CRI updates

```
Update cri-api vendor

update cri-api vendor to include hugepages changes
KEP: kubernetes/enhancements#1199
CRI: kubernetes/kubernetes#83614

Signed-off-by: Byonggon Chun <bg.chun@samsung.com>
```

File: vendor.conf [hunk 2]
```diff
@@ -62,7 +62,7 @@ sigs.k8s.io/yaml fd68e9863619f6ec2fdd8625fe1f02e7c877e480 # v1.1.0
 k8s.io/utils e782cd3c129fc98ee807f3c889c0f26eb7c9daf5
 k8s.io/kubernetes v1.17.1
 k8s.io/klog v1.0.0
-k8s.io/cri-api kubernetes-1.17.1
+k8s.io/cri-api kubernetes-1.18.0-alpha.1
```
Comment (Random-Liu) [inline vendor.conf]: Can you update other kubernetes dependencies correspondingly?

Comment (bg-chun) [reply vendor.conf]: @Random-Liu 
I've updated k8s dependencies to `1.18.0-alpha.1`.

I tried to update dpes to 1.18.0-alpha.2, 
but it causes dependency issue in compile time.
```
# github.com/cri/vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2
vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.pb.go:47:11: undefined: "github.com/cri/vendor/github.com/gogo/protobuf/proto".GoGoProtoPackageIsVersion3
```

I think protobuf version was updated to version3 in kubernetes 1.18-alpha.2.
But, it seems containerd & cri projects use version2 of protobuf.
So It's my plan:
1) asking merging of this PR with k8s 1.18.0-alpha.1 with protobuf v2 vendoring.
-------------------------after k8s 1.18 rel----------------------------
2) update containerd vendors to k8s 1.18 rel & protobuf v3
3) update cri vendors

**Comment by mikebrow** [comment]
hmm odd test fail on only one of the go versions... restarting travis ci

**Comment by mikebrow** [comment]
odd test all won't kick off travis.. and travis is missing the retest button...

**Comment by mikebrow** [comment]
"ERROR: (gcloud.compute.instances.create) Could not fetch resource:
31
 - Quota 'CPUS' exceeded.  Limit: 100.0 in region us-west1."  one of those days...

**Comment by bg-chun** [comment]
humm..... `make test-integration` shows some error.
I will dig it after finishing k8s 1.18 release job...

```
=== RUN   TestVolumeOwnership

--- PASS: TestVolumeOwnership (2.74s)

    volume_copy_up_test.go:98: Create a sandbox

    volume_copy_up_test.go:107: Pull test image

    volume_copy_up_test.go:111: Create a container with volume-ownership test image

    volume_copy_up_test.go:120: Start the container

    volume_copy_up_test.go:125: Check ownership of test directory inside container

    volume_copy_up_test.go:133: Check ownership of test directory on the host

FAIL

./hack/test-utils.sh: line 82: 24626 Terminated              keepalive "sudo PATH=${PATH} ${ROOT}/_output/${CONTAINERD_BIN} ${CONTAINERD_FLAGS}" ${RESTART_WAIT_PERIOD} &> ${report_dir}/containerd.log

Makefile:106: recipe for target 'test-integration' failed

make: *** [test-integration] Error 1

The command "make test-integration" exited with 2.
```

**Comment by mikebrow** [comment]
but works on tip? testing has been so spurious lately... like the host has bad ram or drive sectors.

**Comment by mikebrow** [comment]
closing and re-opening to force a re-test

**Comment by bg-chun** [comment]
> closing and re-opening to force a re-test as it looked to me like travis was just messed up that day

Cool, we've passed all CI jobs. 
Thanks! :)

**Comment by BenTheElder** [comment]
I suspect this PR broke being able to use containerd/cri on systems _without_ hugetlb
https://github.com/kubernetes-sigs/kind/issues/1637

**Pull request merged**